File: 3734-lexicographically-smallest-palindromic-permutation-greater-than-target/3734-lexicographically-smallest-palindromic-permutation-greater-than-target.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    string lexPalindromicPermutation(string s, string target) {
        vector<int> count(26, 0);
        for(char c : s) {
            count[c - 'a']++;
        }
        
        int oddCount = 0;
        string mid_char = "";
        vector<int> half_count(26, 0);
        
        for(int i = 0; i < 26; ++i) {
            if(count[i] % 2 != 0) {
                oddCount++;
                mid_char = string(1, i + 'a');
            }
            half_count[i] = count[i] / 2;
        }
        
        if(oddCount > 1) return "";
        
        int n = s.length();
        int L = n / 2;
        
        for(int i = L; i >= 0; --i) {
            vector<int> pref_count(26, 0);
            bool possible = true;
            
            for(int j = 0; j < i; ++j) {
                pref_count[target[j] - 'a']++;
            }
            
            for(int c = 0; c < 26; ++c) {
                if(pref_count[c] > half_count[c]) {
                    possible = false;
                    break;
                }
            }
            
            if(!possible) continue;
            
            vector<int> rem_count(26, 0);
            for(int c = 0; c < 26; ++c) {
                rem_count[c] = half_count[c] - pref_count[c];
            }
            
            if(i == L) {
                string P = target.substr(0, L);
                string rev_P = P;
                reverse(rev_P.begin(), rev_P.end());
                string pal = P + mid_char + rev_P;
                
                if(pal > target) return pal;
            } else {
                char best_c = 0;
                for(int c = target[i] - 'a' + 1; c < 26; ++c) {
                    if(rem_count[c] > 0) {
                        best_c = c + 'a';
                        rem_count[c]--;
                        break;
                    }
                }
                
                if(best_c != 0) {
                    string P = target.substr(0, i) + best_c;
                    for(int c = 0; c < 26; ++c) {
                        P += string(rem_count[c], c + 'a');
                    }
                    string rev_P = P;
                    reverse(rev_P.begin(), rev_P.end());
                    
                    return P + mid_char + rev_P;
                }
            }
        }
        
        return "";
    }
};

File: 3734-lexicographically-smallest-palindromic-permutation-greater-than-target/3734-lexicographically-smallest-palindromic-permutation-greater-than-target.py
from collections import Counter

class Solution:
    def lexPalindromicPermutation(self, s: str, target: str) -> str:
        counts = Counter(s)
        odd_count = sum(1 for v in counts.values() if v % 2 != 0)

        if odd_count > 1:
            return ""

        mid_char = ""

        for c, v in counts.items():
            if v % 2 != 0:
                mid_char = c

        half_counts = {c: v // 2 for c, v in counts.items()}
        L = len(s) // 2

        for i in range(L, -1, -1):
            pref_counts = Counter(target[:i])

            possible = all(half_counts.get(c, 0) >= v for c, v in pref_counts.items())

            if not possible:
                continue

            rem_counts = {c: half_counts.get(c, 0) - pref_counts.get(c, 0) for c in half_counts}

            if i == L:
                P = target[:L]
                pal = P + mid_char + P[::-1]
                if pal > target:
                    return pal

            else:
                best_c = None
                for c in sorted(rem_counts.keys()):
                    if rem_counts[c] > 0 and c > target[i]:
                        best_c = c
                        break
                
                if best_c:
                    rem_counts[best_c] -= 1
                    P = target[:i] + best_c
                    
                    for c in sorted(rem_counts.keys()):
                        P += c * rem_counts[c]
                        
                    return P + mid_char + P[::-1]
                    
        return ""
